In [2]:
%cd /content

/content


In [ ]:
!git clone https://github.com/DaphneDana/daphne_EDA.git

Cloning into 'daphne_EDA'...


In [3]:
%cd /content/daphne_EDA

[Errno 2] No such file or directory: '/content/daphne_EDA'
/content


In [4]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
%cd /content/drive/MyDrive
!git clone https://github.com/DaphneDana/daphne_EDA.git


/content/drive/MyDrive
Cloning into 'daphne_EDA'...


In [ ]:

%cd /content/drive/MyDrive

# Copy from its current folder into the repo folder
!cp "Colab Notebooks/dana_eda.ipynb" "daphne_EDA/dana_eda.ipynb"


/content/drive/MyDrive


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA

!git status
!git add dana_eda.ipynb
!git commit -m "Add dana_eda notebook"
!git push


/content/drive/MyDrive/daphne_EDA
On branch main

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	dana_eda.ipynb

nothing added to commit but untracked files present (use "git add" to track)
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@9f26df93a605.(none)')
error: src refspec refs/heads/main does not match any
error: failed to push some refs to 'https://github.com/DaphneDana/daphne_EDA.git'


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA #daphne_EDA is my repository name


/content/drive/MyDrive/daphne_EDA


In [7]:
!git config --global user.email "nakamyukadaphne62@gmail.com"
!git config --global user.name "DaphneDana"


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA

!git add dana_eda.ipynb
!git commit -m "Add dana_eda notebook"


/content/drive/MyDrive/daphne_EDA
On branch main
Your branch is based on 'origin/main', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

nothing to commit, working tree clean


In [ ]:
!git status

On branch main
Your branch is based on 'origin/main', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

nothing to commit, working tree clean


In [ ]:
!git push -u origin main


fatal: could not read Username for 'https://github.com': No such device or address


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA

# OPTIONAL: just to see current remote
!git remote -v


/content/drive/MyDrive/daphne_EDA
origin	https://github.com/DaphneDana/daphne_EDA.git (fetch)
origin	https://github.com/DaphneDana/daphne_EDA.git (push)


In [ ]:
import getpass

%cd /content/drive/MyDrive/daphne_EDA

token = getpass.getpass("Enter your GitHub PAT: ")

!git remote set-url origin https://{token}@github.com/DaphneDana/daphne_EDA.git


/content/drive/MyDrive/daphne_EDA
Enter your GitHub PAT: ··········


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA
!git push -u origin main


/content/drive/MyDrive/daphne_EDA
Enumerating objects: 3, done.
Counting objects: 100% (3/3), done.
Delta compression using up to 2 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 1.02 KiB | 65.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/DaphneDana/daphne_EDA.git
 * [new branch]      main -> main
Branch 'main' set up to track remote branch 'main' from 'origin'.


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA


/content/drive/MyDrive/daphne_EDA


In [ ]:
%%bash
cat << 'EOF' > push.sh
#!/bin/bash
# Simple push script for daphne_EDA

# Go to the folder where this script is located
cd "$(dirname "$0")"

# Stage the notebook
git add dana_eda.ipynb

# Use the first argument as commit message, or a default one
MSG=${1:-"Update dana_eda notebook"}

# Commit (if there is something to commit)
git commit -m "$MSG" || echo "Nothing to commit."

# Push to origin main
git push origin main
EOF


In [ ]:
%cd /content/drive/MyDrive/daphne_EDA
!chmod +x push.sh


/content/drive/MyDrive/daphne_EDA


In [ ]:
!unzip /content/uganda_banking_dataset.zip -d /content/

Archive:  /content/uganda_banking_dataset.zip
  inflating: /content/uganda_banking_dataset/ug_bank_customers.csv  
  inflating: /content/uganda_banking_dataset/ug_bank_accounts.csv  
  inflating: /content/uganda_banking_dataset/ug_bank_transactions.csv  
  inflating: /content/uganda_banking_dataset/ug_bank_loans.csv  
  inflating: /content/uganda_banking_dataset/README.md  


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df = pd.read_csv('/content/uganda_banking_dataset/ug_bank_customers.csv' , engine='python', on_bad_lines='skip' )
print (df.info)


<bound method DataFrame.info of      customer_id     district    region  birth_year employment_status  \
0        C100000        Jinja   Eastern        2001     public_sector   
1        C100001      Nansana   Central        1999         sme_owner   
2        C100002       Kabale   Western        1994            farmer   
3        C100003         Lira  Northern        2004            farmer   
4        C100004      Kampala   Central        1979         sme_owner   
...          ...          ...       ...         ...               ...   
2495     C102495  Fort Portal   Western        2003           student   
2496     C102496       Masaka   Central        1974     public_sector   
2497     C102497       Luwero   Central        1985     public_sector   
2498     C102498      Masindi   Western        2001           student   
2499     C102499        Jinja   Eastern        1994     self_employed   

      monthly_income_ugx  kyc_level phone_carrier  has_smartphone date_joined  
0          

In [ ]:
print (df.describe)
print (df.shape)

<bound method NDFrame.describe of      customer_id     district    region  birth_year employment_status  \
0        C100000        Jinja   Eastern        2001     public_sector   
1        C100001      Nansana   Central        1999         sme_owner   
2        C100002       Kabale   Western        1994            farmer   
3        C100003         Lira  Northern        2004            farmer   
4        C100004      Kampala   Central        1979         sme_owner   
...          ...          ...       ...         ...               ...   
2495     C102495  Fort Portal   Western        2003           student   
2496     C102496       Masaka   Central        1974     public_sector   
2497     C102497       Luwero   Central        1985     public_sector   
2498     C102498      Masindi   Western        2001           student   
2499     C102499        Jinja   Eastern        1994     self_employed   

      monthly_income_ugx  kyc_level phone_carrier  has_smartphone date_joined  
0        

In [ ]:
df.head()

customer_id district    region  birth_year employment_status  \
0     C100000    Jinja   Eastern        2001     public_sector   
1     C100001  Nansana   Central        1999         sme_owner   
2     C100002   Kabale   Western        1994            farmer   
3     C100003     Lira  Northern        2004            farmer   
4     C100004  Kampala   Central        1979         sme_owner   

   monthly_income_ugx  kyc_level phone_carrier  has_smartphone date_joined  
0            243537.0        1.0           MTN             1.0  2023-10-14  
1             40526.0        2.0           MTN             1.0  2022-06-19  
2            292956.0        NaN           MTN             1.0  2022-03-23  
3           1102922.0        NaN        Airtel             1.0  2025-08-25  
4            757387.0        2.0        Airtel             1.0  2024-05-06

In [ ]:
df.tail()

customer_id     district   region  birth_year employment_status  \
2495     C102495  Fort Portal  Western        2003           student   
2496     C102496       Masaka  Central        1974     public_sector   
2497     C102497       Luwero  Central        1985     public_sector   
2498     C102498      Masindi  Western        2001           student   
2499     C102499        Jinja  Eastern        1994     self_employed   

      monthly_income_ugx  kyc_level phone_carrier  has_smartphone date_joined  
2495            333980.0        3.0        Airtel             1.0  2025-04-24  
2496            733320.0        2.0      Africell             1.0  2025-07-19  
2497            750008.0        1.0           MTN             1.0  2023-04-05  
2498             66193.0        1.0           MTN             1.0  2022-11-11  
2499            389156.0        3.0        Airtel             1.0  2023-10-12

In [ ]:
df.shape

(2500, 10)

In [ ]:
df.columns

Index(['customer_id', 'district', 'region', 'birth_year', 'employment_status',
       'monthly_income_ugx', 'kyc_level', 'phone_carrier', 'has_smartphone',
       'date_joined'],
      dtype='object')

In [ ]:

# Quick preview of unique values per column
df.nunique()

customer_id           2500
district                60
region                   4
birth_year              48
employment_status        8
monthly_income_ugx    2497
kyc_level                3
phone_carrier            3
has_smartphone           2
date_joined           1174
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   customer_id         2500 non-null   object 
 1   district            2500 non-null   object 
 2   region              2500 non-null   object 
 3   birth_year          2500 non-null   int64  
 4   employment_status   2500 non-null   object 
 5   monthly_income_ugx  2500 non-null   float64
 6   kyc_level           2378 non-null   float64
 7   phone_carrier       2500 non-null   object 
 8   has_smartphone      2381 non-null   float64
 9   date_joined         2500 non-null   object 
dtypes: float64(3), int64(1), object(6)
memory usage: 195.4+ KB


In [ ]:
# Summary stats for numeric columns
df.describe()

birth_year  monthly_income_ugx    kyc_level  has_smartphone
count  2500.000000        2.500000e+03  2378.000000     2381.000000
mean   1989.123200        8.463372e+05     1.639193        0.729525
std      11.465557        2.519276e+06     0.672317        0.444298
min    1958.000000        1.845900e+04     1.000000        0.000000
25%    1981.000000        2.265828e+05     1.000000        0.000000
50%    1991.000000        4.347600e+05     2.000000        1.000000
75%    1999.000000        8.607800e+05     2.000000        1.000000
max    2005.000000        9.165664e+07     3.000000        1.000000

In [ ]:
# Include categorical columns too
df.describe(include='object')

customer_id district   region employment_status phone_carrier  \
count         2500     2500     2500              2500          2500   
unique        2500       60        4                 8             3   
top        C102499  Kampala  Central     self_employed           MTN   
freq             1      464     1285               328          1473   

       date_joined  
count         2500  
unique        1174  
top     2023-10-11  
freq             7

📊 Summary of Initial Data Inspection

This section provides an overview and interpretation of the descriptive statistics generated from the dataset. The dataset contains 2,500 bank customer records with demographic information, financial attributes, and account onboarding details.

## 1. Summary of Numerical Variables
### Birth Year

All 2,500 entries have valid birth years.

The average birth year is 1989, meaning the typical customer is around 36 years old.

Ages range from approximately 20 years (born 2005) to 67 years (born 1958).

The distribution appears realistic and well-behaved with no implausible values.

Implication:
The age range is diverse but valid, providing good demographic coverage without requiring corrections.

### Monthly Income (UGX)

All entries contain income values.

The average monthly income is UGX 846,337, with a median of UGX 434,760, indicating right-skewness.

The standard deviation (UGX 2.5M) is extremely high, suggesting significant income variability.

The maximum income (UGX 91M) is an extreme outlier, far above typical Ugandan income ranges.

Implication:
Income data is highly skewed with substantial outliers. Outlier detection and possible capping or transformation will be necessary during preprocessing.

### KYC Level

Out of 2,500 entries, 2,378 have valid KYC values, meaning 122 are missing.

The average KYC level is 1.64, with valid levels ranging from 1 to 3.

Distribution suggests most customers fall under KYC levels 1 and 2.

Implication:
KYC values are straightforward but missing values require imputation (likely using the median value).

### Smartphone Ownership

2,381 customers have valid smartphone status values → 119 missing.

73% of customers own a smartphone.

Binary distribution (0/1) is clean and interpretable.

Implication:
Missing values will need imputation. The mode (1) is most appropriate since smartphone ownership is common.

## 2. Summary of Categorical Variables
### Customer ID

All 2,500 IDs are unique, confirming no duplicate customer records.

### District

Data spans 60 districts, with Kampala having the highest representation (464 customers).

Implication:
Strong urban concentration, especially around Kampala.

### Region

Four valid regions appear: Central, Eastern, Western, Northern.

Central region dominates with 1,285 customers (51%).

Implication:
The dataset may reflect customer concentration in more urbanized regions.

### Employment Status

There are 8 employment categories, with self-employed being the most common (328 customers).

Implication:
Employment categories appear reasonable but may require text normalization for consistency.

### Phone Carrier

Only 3 carriers appear (MTN, Airtel, Africell), with MTN being most common (59%).

Implication:
Carrier distribution aligns with Uganda’s market share. Minor text cleaning may be necessary.

### Date Joined

Contains 1,174 unique joining dates, with some dates repeated up to 7 times.

Implication:
Column looks structurally correct but must be converted to actual datetime. Future dates should be validated.

## 3. Key Issues Identified for Cleaning

Missing Values

KYC Level (122 missing)

Smartphone Ownership (119 missing)

Outliers

Monthly income has extremely high values (e.g., 91M UGX).

Data Type Fixes Needed

date_joined → convert to datetime

has_smartphone → convert to boolean/int

kyc_level, birth_year, monthly_income_ugx → ensure numeric

Categorical Cleanup

Standardize text values (e.g., region names, phone carriers, employment statuses)

Possible Feature Engineering

Age (from birth_year)

Customer tenure (from date_joined)

## 4. Overall Interpretation

The dataset is generally well-structured and rich in useful customer-level information. Most numeric and categorical values fall within realistic ranges, and customer IDs are completely unique. However, several columns require additional cleaning, mainly around missing values, outlier handling, and type conversion. Addressing these issues will prepare the dataset for meaningful exploratory analysis and profiling.

In [ ]:
%cd /content/drive/MyDrive/daphne_EDA
!./push.sh "Made data analyzing"

/content/drive/MyDrive/daphne_EDA
On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	push.sh

nothing added to commit but untracked files present (use "git add" to track)
Nothing to commit.
Everything up-to-date


In [ ]:
!git add .



In [ ]:
!git commit -m "Made some data anaysis"

[main b4dee28] Made some data anaysis
 1 file changed, 17 insertions(+)
 create mode 100755 push.sh


In [ ]:
!git push origin main

Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 2 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 557 bytes | 50.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/DaphneDana/daphne_EDA.git
   827a6f1..b4dee28  main -> main


In [ ]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [ ]:
# Count of missing values per column
df.isna().sum()

customer_id             0
district                0
region                  0
birth_year              0
employment_status       0
monthly_income_ugx      0
kyc_level             122
phone_carrier           0
has_smartphone        119
date_joined             0
dtype: int64

In [ ]:
# Percentage of missing values per column
(df.isna().sum() / len(df)) * 100

customer_id           0.00
district              0.00
region                0.00
birth_year            0.00
employment_status     0.00
monthly_income_ugx    0.00
kyc_level             4.88
phone_carrier         0.00
has_smartphone        4.76
date_joined           0.00
dtype: float64

In [ ]:
missing_summary = (
    df.isna().sum()
      .to_frame(name='missing_count')
      .assign(missing_pct=lambda x: x['missing_count'] * 100 / len(df))
)
missing_summary


missing_count  missing_pct
customer_id                     0         0.00
district                        0         0.00
region                          0         0.00
birth_year                      0         0.00
employment_status               0         0.00
monthly_income_ugx              0         0.00
kyc_level                     122         4.88
phone_carrier                   0         0.00
has_smartphone                119         4.76
date_joined                     0         0.00

In [ ]:
# Impute KYC level with median
df['kyc_level'] = df['kyc_level'].fillna(df['kyc_level'].median())


In [ ]:
# Impute smartphone ownership with mode
df['has_smartphone'] = df['has_smartphone'].fillna(df['has_smartphone'].mode()[0])


In [5]:
%cd /content/drive/MyDrive

!cp "Colab Notebooks/dana_eda.ipynb" "daphne_EDA/dana_eda.ipynb"


/content/drive/MyDrive


In [8]:
%cd /content/drive/MyDrive/daphne_EDA

!git add dana_eda.ipynb
!git commit -m "Sync latest notebook from Colab Notebooks"
!git push


/content/drive/MyDrive/daphne_EDA
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   push.sh

no changes added to commit (use "git add" and/or "git commit -a")
Everything up-to-date


In [9]:
!git add .

In [10]:
!git commit -m "Sync latest notebook from Colab Notebooks"

[main db955e1] Sync latest notebook from Colab Notebooks
 1 file changed, 0 insertions(+), 0 deletions(-)
 mode change 100755 => 100644 push.sh


In [11]:
!git push origin main

Enumerating objects: 3, done.
Counting objects: 100% (3/3), done.
Delta compression using up to 2 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 298 bytes | 59.00 KiB/s, done.
Total 2 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/DaphneDana/daphne_EDA.git
   b4dee28..db955e1  main -> main


In [12]:
%cd /content/drive/MyDrive

# Show both locations just to be sure
!ls "Colab Notebooks"
!ls daphne_EDA

# Now overwrite the repo notebook with the good one
!cp "Colab Notebooks/dana_eda.ipynb" "daphne_EDA/dana_eda.ipynb"


/content/drive/MyDrive
'01_wrangling (1).ipynb'
 01_wrangling.ipynb
'Another copy of dana_eda.ipynb'
 audio_files.ipynb
'Copy of Acquiring_and_Documenting_Data_Interactive_v2.ipynb'
'Copy of dana_eda.ipynb'
'Copy of EDA_Part2_Relationships_and_Quality.ipynb'
'Copy of Exploring_the_World_of_GenAI_11_25 (1).ipynb'
'Copy of Exploring_the_World_of_GenAI_11_25.ipynb'
'Copy of Supervised Learning (Part 1).ipynb'
'Copy of Wrangling_and_EDA_Module.ipynb'
 dana_eda.ipynb
'GROUP4__PART_1 (1).ipynb'
'GROUP4_ PART 1.ipynb'
'GROUP4_PART2&3 (1).ipynb'
'GROUP4_PART2&3.ipynb'
 Practice1.ipynb
 Untitled
 Untitled0.ipynb
 Untitled1.ipynb
 Untitled2.ipynb
 Untitled3.ipynb
 Untitled4.ipynb
 Untitled5.ipynb
dana_eda.ipynb	push.sh


In [13]:
%cd /content/drive/MyDrive/daphne_EDA
!git status


/content/drive/MyDrive/daphne_EDA
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [14]:
%cd /content/drive/MyDrive

# List all notebooks in Colab Notebooks with sizes
!ls -l "Colab Notebooks"

# List all notebooks in your repo folder with sizes
!ls -l daphne_EDA


/content/drive/MyDrive
total 20779
-rw------- 1 root root    6974 Nov 26 16:43 '01_wrangling (1).ipynb'
-rw------- 1 root root    7491 Nov 26 16:44  01_wrangling.ipynb
-rw------- 1 root root    3046 Nov 28 15:24 'Another copy of dana_eda.ipynb'
-rw------- 1 root root 1486802 Jan  6  2025  audio_files.ipynb
-rw------- 1 root root  196789 Nov 17 05:17 'Copy of Acquiring_and_Documenting_Data_Interactive_v2.ipynb'
-rw------- 1 root root  104708 Dec  2 09:36 'Copy of dana_eda.ipynb'
-rw------- 1 root root  635583 Nov 20 04:58 'Copy of EDA_Part2_Relationships_and_Quality.ipynb'
-rw------- 1 root root 6563236 Nov 12 11:43 'Copy of Exploring_the_World_of_GenAI_11_25 (1).ipynb'
-rw------- 1 root root 5771632 Nov 12 11:49 'Copy of Exploring_the_World_of_GenAI_11_25.ipynb'
-rw------- 1 root root   39512 Nov 25 05:25 'Copy of Supervised Learning (Part 1).ipynb'
-rw------- 1 root root 5467805 Nov 19 05:04 'Copy of Wrangling_and_EDA_Module.ipynb'
-rw------- 1 root root    2265 Nov 28 15:11  dana_eda

In [15]:
%cd /content/drive/MyDrive
!find . -iname "dana_eda*.ipynb" -ls


/content/drive/MyDrive
       35      3 -rw-------   1 root     root         2265 Nov 28 15:11 ./Colab\ Notebooks/dana_eda.ipynb
       29      3 -rw-------   1 root     root         2265 Dec  2 09:35 ./daphne_EDA/dana_eda.ipynb


In [16]:
%cd /content/drive/MyDrive

# Overwrite the repo file with the REAL notebook
!cp "Copy of dana_eda.ipynb" "daphne_EDA/dana_eda.ipynb"

# Check size now inside the repo
%cd /content/drive/MyDrive/daphne_EDA
!ls -l dana_eda.ipynb


/content/drive/MyDrive
cp: cannot stat 'Copy of dana_eda.ipynb': No such file or directory
/content/drive/MyDrive/daphne_EDA
-rw------- 1 root root 2265 Dec  2 09:35 dana_eda.ipynb
